docs/design.md

```diff
@@ -0,0 +1,116 @@
+Design
+======
+
+The main goals of libnetwork are highlighted in the [roadmap](../ROADMAP.md).
+This document describes how libnetwork has been designed in order to acheive this.
+Requirements for individual releases can be found on the [Project Page](https://github.com/docker/libnetwork/wiki)
+
+## Legacy Docker Networking
+
+Prior to libnetwork a container's networking was handled in both Docker Engine and libcontainer.
+Docker Engine was responsible for providing the configuration of the container's networking stack.
+Libcontainer would then use this information to create the necessary networking devices and move them in to a network namespace.
+This namespace would then be used when the container is started.
+
+## The Container Network Model
+
+Libnetwork implements Container Network Model (CNM) which formalizes the steps required to provide networking for containers while providing an abstraction that can be used to support multiple network drivers. The CNM is built on 3 main components.
+
+**Sandbox**
+
+A Sandbox contains the configuration of a container's network stack.
+This includes management of the container's interfaces, routing table and DNS settings. 
+An implementation of a Sandbox could be a Linux Network Namespace, a FreeBSD Jail or other similar concept.
+A Sandbox may contain *many* endpoints from *multiple* networks
+
+**Endpoint**
+
+An Endpoint joins a Sandbox to a Network.
+An implementation of an Endpoint could be a `veth` pair, an Open vSwitch internal port or similar.
+An Endpoint can belong to *only one* network but may only belong to *one* Sandbox
+
+**Network**
+
+A Network is a group of Endpoints that are able to communicate with each-other directly.
+An implementation of a Network could be a Linux bridge, a VLAN etc...
+Networks consist of *many* endpoints
+
+## API
+
+Consumers of the CNM, like Docker for example, interact through the following APIs
+
+The `NetworkController` object is created to manage the allocation of Networks and the binding of these Networks to a specific Driver
+Once a Network is created, `network.CreateEndpoint` can be called to create a new Endpoint in a given network.
+When an Endpoint exists, it can be joined to a Sandbox using `endpoint.Join(id)`. If no Sandbox exists, one will be created, but if the Sandbox already exists, the endpoint will be added there.
+The result of the Join operation is a Sandbox Key which identifies the Sandbox to the Operating System (e.g a path)
+This Key can be passed to the container runtime so the Sandbox is used when the container is started.
+
+When the container is stopped, `endpoint.Leave` will be called on each endpoint within the Sandbox
+Finally once, endpoint.
+
+## Component Lifecycle
+
+### Sandbox Lifecycle
+
+The Sandbox is created on `endpoint.Join` and deleted when `endpoint.Leave` is called on the last endpoint.
```

Comment:
I second this. This is more precise although I would say `The sandbox is created during first endpoint.Join`